### PR TITLE
Add engine_from_config to sqlalchemy integration

### PR DESCRIPTION
**Pull request by pacejackson**
This allows us to specify the different connection params and fetch the password from a SecretsStore.

👓 @spladug @bsimpson63

### CODE REVIEW

File: baseplate/context/sqlalchemy.py [hunk 1]
```diff
@@ -3,14 +3,70 @@
...
 from sqlalchemy.orm import Session
 
-from ..context import ContextFactory
-from ..core import (
-    ServerSpan,
-    SpanObserver,
-)
+from baseplate import config
+from baseplate.context import ContextFactory
+from baseplate.core import ServerSpan, SpanObserver
```
Comment (pacejackson) [inline baseplate/context/sqlalchemy.py]: @spladug how do you feel about me cleaning up these imports while I'm already in here?

Comment (spladug) [reply baseplate/context/sqlalchemy.py]: i feel good. very good.

File: baseplate/context/sqlalchemy.py [hunk 2]
```diff
@@ -3,14 +3,70 @@
...
+    * ``database`` (optional): the database name.
+    * ``query`` (optional): a dictionary of options to be passed to the dialect
+        and/or the DBAPI upon connect.
+
+    """
+    assert prefix.endswith(".")
+    config_prefix = prefix[:-1]
+    cfg = config.parse_config(app_config, {
+        config_prefix: {
+            "drivername": config.String,
```
Comment (spladug) [inline baseplate/context/sqlalchemy.py]: why not allow the url format?

Comment (pacejackson) [reply baseplate/context/sqlalchemy.py]: I kinda feel like that might get confusing about what you are supposed to set?  You can still use `engine_from_config` straight from sqlalchemy if you want which just uses the URL but then you have to embed the password in the URL.

**Comment by pacejackson** [comment]
💇‍♂️ 
I changed it so you can either use `database.url` or `database.drivername`.  If you set the url, it just uses that, it does no go so far to see if you set anything else and raise an error if you do.

As a result, I also moved to make `secrets` an optional argument unless you actually set `password_secret`.

LMK if you think this makes sense or if it's too many conditional requirements (which was my initial concern).

**Comment by spladug** [comment]
What about parsing the URL first then allowing any other config fields to override then kwargs, in that order. That seems like it'd make the most sense (since the common case would be url + password from secret).

File: baseplate/context/sqlalchemy.py [hunk 3]
```diff
@@ -3,14 +3,88 @@
...
+
+    """
+    assert prefix.endswith(".")
+    config_prefix = prefix[:-1]
+    cfg = config.parse_config(app_config, {
+        config_prefix: {
+            "url": config.Optional(config.String),
+            "drivername": config.Optional(config.String),
+            "username": config.Optional(config.String),
+            "password_secret": config.Optional(config.String),
```
Comment (spladug) [inline baseplate/context/sqlalchemy.py]: I think we probably want to have username be in here too. If the credentials are being rotated, that'd generally be done by changing user to a new one with a new password, rolling everyone off the old one, then killing the old user. Right?

Comment (spladug) [reply baseplate/context/sqlalchemy.py]: (do we want to add a "credentials" secret type?)

Comment (pacejackson) [reply baseplate/context/sqlalchemy.py]: 🤔

Comment (pacejackson) [reply baseplate/context/sqlalchemy.py]: https://github.com/reddit/baseplate.py/pull/267

**Comment by pacejackson** [comment]
> What about parsing the URL first then allowing any other config fields to override then kwargs, in that order. That seems like it'd make the most sense (since the common case would be url + password from secret).

That seems like it would get really confusing really quickly other than the case of overwriting the username/password?

**Comment by pacejackson** [comment]
Maybe the thing to do is to _only_ support url + optional credentials? @spladug

**Comment by pacejackson** [comment]
💇‍♂️ 
This simplifies it to only support `url` and `credentials` (note the tests will fail because I'm starting to use code from #267 ).

**Comment by bsimpson63** [comment]
🐟 this looks great

**Comment by pacejackson** [comment]
💇‍♂️ 

Screen Shot 2019-04-10 at 5 14 49 PM

**Pull request merged**